### PR TITLE
Implemented get_output_channel() fn for SocketIO Channel

### PR DESCRIPTION
**Pull request by alfredfrancis**
**Proposed changes**:
- Implemented get_output_channel() fn for SocketIO Channel
- Changed Socketio Output message handlers to use recepient_id instead of self.sid 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @erohmensing will take a look at it as soon as possible ✨

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @degiz will take a look at it as soon as possible ✨

**Comment by chkoss** [comment]
Here is some context as to why this feature was implemented: https://forum.rasa.com/t/integrate-rasa-with-live-chat/22314/6

File: rasa/core/channels/socketio.py [hunk 1]
```diff
@@ -155,6 +160,9 @@ def blueprint(
             sio, self.socketio_path, "socketio_webhook", __name__
         )
 
+        # make sio object static to use in get_output_channel
+        SocketIOInput.sio = sio
```
Comment (degiz) [inline rasa/core/channels/socketio.py:164]: IMO we have only one instance of SocketIOInput created, so that doesn't have to be static. 
I think you can use `self.sio` to store it.

File: rasa/core/channels/socketio.py [hunk 2]
```diff
@@ -145,6 +147,9 @@ def __init__(
         self.namespace = namespace
         self.socketio_path = socketio_path
 
+    def get_output_channel(self) -> Optional["OutputChannel"]:
+        return SocketIOOutput(SocketIOInput.sio , None, self.bot_message_evt)
```
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/channels/socketio.py:151]: And please see comment below about `SocketIOInput.sio`

Comment (wochinge) [inline rasa/core/channels/socketio.py:151]: I think we should handle the case where `SocketIOInput.sio` is `None` (e.g. if the Rasa Open Server uses multiple Sanic workers (which equals multiple processes))

File: rasa/core/channels/socketio.py [hunk 3]
```diff
@@ -95,27 +95,29 @@ async def send_elements(
...
     ) -> None:
         """Sends an attachment to the user."""
-        await self._send_message(self.sid, {"attachment": attachment})
+        await self._send_message(recipient_id, {"attachment": attachment})
 
 
 class SocketIOInput(InputChannel):
     """A socket.io input channel."""
 
+    sio = None
```
Comment (wochinge) [inline rasa/core/channels/socketio.py:119]: Mhm, I'm a bit torn to be honest. Yes it works now as we are only using one Sanic worker (= 1 process). As soon as we are using multiple Sanic workers or multiple Rasa Open Source instances, this will break (meaning that the message is lost).

File: rasa/core/channels/socketio.py [hunk 4]
```diff
@@ -95,27 +95,29 @@ async def send_elements(
...
+        json_message.setdefault("room", recipient_id)
 
         await self.sio.emit(self.bot_message_evt, **json_message)
 
     async def send_attachment(
         self, recipient_id: Text, attachment: Dict[Text, Any], **kwargs: Any
     ) -> None:
         """Sends an attachment to the user."""
-        await self._send_message(self.sid, {"attachment": attachment})
+        await self._send_message(recipient_id, {"attachment": attachment})
```
[the remark this answers is not in the document]
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:113]: We can join the user to a room name = session_id to address this case.

```
@sio.on("session_request", namespace=self.namespace)
async def session_request(sid: Text, data: Optional[Dict]):
    if data is None:
        data = {}
    if "session_id" not in data or data["session_id"] is None:
        data["session_id"] = uuid.uuid4().hex
    if self.session_persistence:
        sio.enter_room(sid,data["session_id"])
    await sio.emit("session_confirm", data["session_id"], room=sid)
    logger.debug(f"User {sid} connected to socketIO endpoint.")
```
Shall I proceed with this change?

File: rasa/core/channels/socketio.py (hunk 2 above)
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:151]: Any suggestions on how to do it?

File: rasa/core/channels/socketio.py (hunk 1 above)
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:164]: What about when we use multiple Sanic workers?

File: rasa/core/channels/socketio.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/core/channels/socketio.py:113]: Great idea with `enter_room`. I think the change might make more sense here:

```python
 if self.session_persistence:
                if not data.get("session_id"):
                    raise_warning(
                        "A message without a valid session_id "
                        "was received. This message will be "
                        "ignored. Make sure to set a proper "
                        "session id using the "
                        "`session_request` socketIO event."
                    )
                    return
                 sio.enter_room(sid,data["session_id"]). # <-- replaced this line

           sender_id = sid # <-- always do this
```

File: rasa/core/channels/socketio.py (hunk 2 above)
Comment (wochinge) [reply rasa/core/channels/socketio.py:151]: How about printing a warning and returning `None`?

File: rasa/core/channels/socketio.py (hunk 1 above)
Comment (wochinge) [reply rasa/core/channels/socketio.py:164]: Shouldn't matter. Each sanic worker uses a different process which means that they don't share any (not process safe) variables anyway.

File: rasa/core/channels/socketio.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:113]: Yeah, that's better.

[the remark this answers is not in the document]
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:113]: if we write the **enter_room** in **handle_message** it'll get executed for each message whereas only once in **session_request**

[the remark this answers is not in the document]
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:113]: Also, if the session_persistence is true, shouldn't sender_id be equal to session_id?

File: rasa/core/channels/socketio.py (hunk 2 above)
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py]: @wochinge Updated with suggetions

File: rasa/core/channels/socketio.py (hunk 1 above)
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py]: I've updated pull request with the suggestion

**Comment by m-vdb** [comment]
@wochinge @alfredfrancis 💡 Rasa Open Source release is scheduled for next Tuesday April 28th.  Any chance we can ship this as part of the milestone?

**Comment by alfredfrancis** [comment]
> @wochinge @alfredfrancis  Rasa Open Source release is scheduled for next Tuesday April 28th. Any chance we can ship this as part of the milestone?

I'm waiting for the comments from @wochinge

**Comment by wochinge** [comment]
@alfredfrancis I have to do another review first, then I'm back on this 👍

File: rasa/core/channels/socketio.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/core/channels/socketio.py:112]: You're completely right 👍 Sorry!

File: rasa/core/channels/socketio.py [hunk 5]
```diff
@@ -28,9 +28,8 @@ class SocketIOOutput(OutputChannel):
     def name(cls) -> Text:
         return "socketio"
 
-    def __init__(self, sio, sid, bot_message_evt) -> None:
+    def __init__(self, sio, bot_message_evt) -> None:
```
Comment (wochinge) [inline rasa/core/channels/socketio.py:31]:
```suggestion
    def __init__(self, sio: AsyncServer, bot_message_evt: Text) -> None:
```

File: rasa/core/channels/socketio.py [hunk 6]
```diff
@@ -144,6 +143,17 @@ def __init__(
...
         self.namespace = namespace
         self.socketio_path = socketio_path
+        self.sio = None
+
+    def get_output_channel(self) -> Optional["OutputChannel"]:
+        if self.sio == None:
+            raise_warning(
+                "Output channel can't be"
+                "reproduced in case if sio equals to None"
+                "e.g. if the Rasa Server uses multiple Sanic workers"
```
Comment (wochinge) [inline rasa/core/channels/socketio.py:153]: I'd rephrase it too be a bit less technical (currently you'd need to know the code to understand what `sio` means).

What do you think about this one?
```suggestion
                "SocketIO output channel cannot be recreated. "
                "This is expected behavior when using multiple Sanic "
                "workers or multiple Rasa Open Source instances. "
                "Please use a different channel for external events in these "
                "scenarios."
```

File: rasa/core/channels/socketio.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py:112]: No problemo +1

**Comment by alfredfrancis** [comment]
> Great work @alfredcogniwide  Left two small comments, but otherwise ready to go.
> Can you please make sure that the CI tests here pass https://github.com/RasaHQ/rasa/actions/runs/81645737 ? If any results there are wrong, please let me know. We / GitHub recently had some problems with GitHub actions

Ok, let me try

File: rasa/core/channels/socketio.py (hunk 6 above)
Comment (alfredfrancis) [reply rasa/core/channels/socketio.py]: This is much better 👍

**Comment by wochinge** [comment]
@alfredfrancis Could you please also sign the CLA? Thanks 🎉

**Comment by alfredcogniwide** [comment]
> @alfredfrancis Could you please also sign the CLA? Thanks

Done

**Comment by alfredfrancis** [comment]
> @alfredfrancis Can you please also add a changelog entry?

Could you help me do this ? Shall ? add an entry under [1.9.7] - 2020-04-23

**Comment by wochinge** [comment]
Awesome! I quickly added changelog and a small docs change, because we want to include this in the release today 🎉

**Comment by alfredfrancis** [comment]
> Awesome! I quickly added changelog and a small docs change, because we want to include this in the release today

Thank you

**Comment by alfredfrancis** [comment]
> Thanks for the contribution @alfredfrancis

You are most welcome :heart:
